mat64/dense_arithmetic.go

```diff
@@ -361,9 +362,12 @@ func (m *Dense) Mul(a, b Matrix) {
 		panic(ErrShape)
 	}
 
-	var w Dense
+	var w *Dense
 	if m != a && m != b {
-		w = *m
+		w = m
+	} else {
+		m.reuseAs(ar, bc)
```

Comment:
This shouldn't be necessary. Since m is the same as an input argument, it can only be ar x ac, or br x bc. Either the multiplication ends with that size, or the result should panic because the receiver is non-zero and the dimensions do not match. While the code you added in the inner test looks deceptively nice, I think it's more surprising to overwrite a mis-sized receiver.